modulus/models/module.py

```diff
@@ -65,55 +106,286 @@ def debug(self):
+            )
+        if not directory.is_dir():
+            os.makedirs(directory)
+
+        torch.save(self.state_dict(), directory.joinpath("model.pt"))
+
+        with open(directory.joinpath("args.json"), "w") as f:
+            json.dump(self._args, f)
+            if file_name is None:
+                file_name = self.meta.name + ".pt"
+
+        # Save the modulus version and git hash (if available)
+        metadata_info = {"modulus_version": modulus.__version__,
+                         "mdlus_file_version": self.__version__}
+
+        if save_git_hash:
+            import git
+            repo = git.Repo(search_parent_directories=True)
+            try:
+                metadata_info["git_hash"] = repo.head.object.hexsha
+            except git.InvalidGitRepositoryError:
+                metadata_info["git_hash"] = None
+
+        with open(directory.joinpath("metadata.json"), "w") as f:
+            json.dump(metadata_info, f)
+
+    @staticmethod
+    def _check_checkpoint(file_name: str) -> bool:
+        if not file_name.endswith(Module._file_extension):
+            raise ValueError(
+                f"File name must end with {Module._file_extension} extension"
+            )
+
         if file_name is None:
-            file_name = self.meta.name + ".pt"
+            raise ValueError("File name must be provided to load the model")
+
+        directory = Path(file_name)
+        if not directory.is_dir():
+            raise IOError(f"Model directory {directory} not found")
+
+        if not directory.joinpath("args.json").exists():
+            raise IOError(
+                f"Model checkpoint {directory.joinpath('model.json')} not found"
+            )
 
-        file_name = Path(file_name)
-        if not file_name.parents[0].is_dir():
+        if not directory.joinpath("metadata.json").exists():
             raise IOError(
-                f"Model checkpoint parent directory {file_name.parents[0]} not found"
+                f"Model checkpoint {directory.joinpath('metadata.json')} not found"
             )
 
-        torch.save(self.state_dict(), str(file_name))
+        if not directory.joinpath("model.pt").exists():
+            raise IOError(
+                f"Model checkpoint {directory.joinpath('model.pt')} not found"
+            )
 
-    def load(self, file_name: Union[str, None] = None) -> None:
-        """Simple utility for loading the model from checkpoint
+        # Check if the checkpoint version is compatible with the current version
+        with open(directory.joinpath("metadata.json"), "r") as f:
+            metadata_info = json.load(f)
+            if metadata_info['mdlus_file_version'] != Module.__version__:
+                raise IOError(
+                    f"Model checkpoint version {metadata_info['mdlus_file_version']} is not compatible with current version {Module.__version__}"
+                )
+
+        return directory
```

Comment:
Why is the return type bool but a Path object is being returned? Also, wouldn't we want all of the `raise`s to be actually `return False`? Or maybe change the return type to an enum and return either success or an error code?